### PR TITLE
feat(http): detect complex schema error, improve error message

### PR DESCRIPTION
**Pull request by daniel-white**
Addresses #2033

**Summary**

* upgrade json-schema-sampler to get better error type
* rejigger to allow prism specific error to be thrown

**Checklist**

- The basics
  - [x] I tested these changes manually in my local or dev environment
- Tests
  - [x] Added or updated
  - [ ] N/A
- Event Tracking
  - [ ] I added event tracking and followed the event tracking guidelines
  - [ ] N/A
- Error Reporting
  - [ ] I reported errors and followed the error reporting guidelines
  - [ ] N/A

**Screenshots**

If applicable, add screenshots or gifs to help demonstrate the changes. If not applicable, remove this screenshots
section before creating the PR.

**Additional context**

Add any other context about the pull request here. Remove this section if there is no additional context.

### CODE REVIEW

File: packages/http/src/mocker/generator/JSONSchema.ts [hunk 1]
```diff
@@ -98,6 +102,26 @@ export function sortSchemaAlphabetically(source: any): any {
...
+}
+
+export class GeneratorError extends Error {}
+
+export class SchemaTooComplexGeneratorError extends GeneratorError {
+  constructor(operation: IHttpOperation, public readonly cause: Error) {
+    super(
+      `The operation ${operation.method.toUpperCase()} ${
+        operation.path
+      } references a JSON Schema that is too complex to generate.`
```
Comment (EdVinyard) [inline packages/http/src/mocker/generator/JSONSchema.ts:124]: Is there a way to state precisely in what way/dimension the spec is too complex here?  As a Prism user who sees this, what can I do to work-around the problem or simplify the spec to make it fall within the boundary?

**Review by EdVinyard — CHANGES_REQUESTED** [review]
Everything else looks good here, but I'd love to see a really specific error message that suggests a work-around or fix for users.  I'd be happy to brainstorm that with you, or maybe even bring in our docs folks to wordsmith it once we can nail the description of what limit/boundary you've surpassed.

File: packages/http/src/mocker/generator/JSONSchema.ts (hunk 1 above)
Comment (daniel-white) [reply packages/http/src/mocker/generator/JSONSchema.ts:124]: we'd have to make json-schema-sampler and json-schema-faker to provide that context. the sampler code is all vanilla js so its pretty terse to follow

Comment (EdVinyard) [reply packages/http/src/mocker/generator/JSONSchema.ts:124]: Is this a case where we could add to the Prism docs we publish, and include a URL here for more info?  I know we haven't done this before, but I'm just trying to find _some_ way to give users information they can act on (or at least understand).  I just don't think, as a Prism user, I'd be able to understand "too complex to generate" enough to solve my own problem.

**Review by EdVinyard — APPROVED** [review]

File: docs/guides/06-errors.md [hunk 2]
```diff
@@ -264,6 +264,20 @@ paths:
...
 **Explanation:** This error occurs when the current request specifies a `Content-Type` that isn't supported by corresponding HTTP Operation. In the case there is no request body or `Content-Length` is 0, the `Content-Type` header is ignored.
 
+
+### SCHEMA_TOO_COMPLEX
+
+**Message: Unable to generate [body|header] for response. The schema is too complex to generate.**
+
+**Returned Status Code: `500`**
+
+**Explanation:** This error occurs when part of the response can't be generated using JSON Schema Sampler. JSON Schema Sampler has been configured to use a limit of 2500 "ticks". A "tick" is loosely defined as any instance of a JSON Schema schema or subschema. This includes a/an single:
```
Comment (RaySalty) [inline docs/guides/06-errors.md:274]: Recommend: "This error occurs when part of the response can't be generated using **the** JSON Schema Sampler. **The** JSON Schema Sample **has** a limit of 2500 ticks."

"Has been configured" makes it feel open to changing.

File: docs/guides/06-errors.md [hunk 3]
```diff
@@ -264,6 +264,20 @@ paths:
...
+* Property
+* Object
+* Array item (each in an array between `minItems` and `maxItems`, defaulting to 1)
+* Combiner item (members of `allOf`, `oneOf`, `anyOf`)
+ 
+...all recursively, unless a `$ref` has already been visited (the subschema is skipped). As a workaround, consider using [dynamic response mocking](./11-dynamic-response-with-faker.md).
```
Comment (RaySalty) [inline docs/guides/06-errors.md:280]: If "(**in which case**, the subschema is skipped)" is a correct understanding of the use of `$ref`, then recommend adding the additional language to read even easier.

**Review by RaySalty — APPROVED** [review]

File: docs/guides/06-errors.md (hunk 2 above)
Comment (daniel-white) [reply docs/guides/06-errors.md:274]: it is open to changing

**Pull request merged**